graph-http/src/download/client.rs

```diff
@@ -284,76 +291,106 @@ impl AsyncDownload {
-            }
-
-            self.client.set_request(vec![
-                RequestAttribute::ClearHeaders,
-                RequestAttribute::Method(Method::GET),
-                RequestAttribute::RequestType(RequestType::Basic),
-                RequestAttribute::Url(GraphUrl::from(response.url().clone())),
-            ])?;
+        if self.client.request_type() == RequestType::Redirect {
+            let response = self
+                .client
+                .build()
+                .await
+                .send()
+                .await?
+                .with_graph_error()
+                .await?;
+
+            let mut client = self.client.client.lock().await;
+            (*client).headers.clear();
+            (*client).method = Method::GET;
+            (*client).req_type = RequestType::Basic;
+            (*client).url = GraphUrl::from(response.url().clone());
         }
 
-        let response = self.client.build().await.send().await?;
-        if let Ok(mut error) = GraphError::try_from(&response) {
-            let error_message: GraphResult<ErrorMessage> =
-                response.json().await.map_err(GraphFailure::from);
-            if let Ok(message) = error_message {
-                error.set_error_message(message);
-            }
-            return Err(GraphFailure::from(error));
-        }
+        let response = self
+            .client
+            .build()
+            .await
+            .send()
+            .await?
+            .with_graph_error()
+            .await?;
 
         let path = {
-            if let Some(name) = request.file_name.as_ref() {
-                self.check_file_name_length(name)?;
-                request.path.join(name)
-            } else if let Some(name) = self.parse_content_disposition(response.headers()) {
-                self.check_file_name_length(&name)?;
+            if let Some(name) = request
+                .file_name
+                .clone()
+                .or_else(|| self.parse_content_disposition(response.headers()))
+            {
+                if name.len() > MAX_FILE_NAME_LEN {
+                    return Err(AsyncDownloadError::FileNameTooLong);
+                }
                 request.path.join(name)
             } else {
-                return GraphRsError::DownloadFileName.as_err_res();
+                return Err(AsyncDownloadError::NoFileName);
             }
         };
 
         if let Some(ext) = request.extension.as_ref() {
             path.with_extension(ext.as_str());
         }
 
-        self.check_existing_file(&path).await?;
-        IoTools::copy_async(path, response).await
+        if path.exists() && !self.is_overwrite_existing_file().await {
+            return Err(AsyncDownloadError::FileExists(
+                path.to_string_lossy().to_string(),
+            ));
+        }
+
+        Ok(iotools::copy_async(path, response).await?)
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum AsyncDownloadError {
```

Comment:
```suggestion
#[allow(clippy::large_enum_variant)]
pub enum AsyncDownloadError {
```
This is needed on both the blocking download and async download or clippy will throw a fit for a large enum variant. I'm not worried about this lint, at least for now, because it doesn't have a performance impact.